### PR TITLE
Add: [CDN] code and config to generate required files for CDN

### PR DESCRIPTION
**Pull request by TrueBrain**
This includes:
 - index.html: an index file (as S3 / CloudFront / CloudFlare
               cannot do this for us)
 - folders.yaml: to be able to script walking the CDN
 - latest.yaml: a short summary for the latest release of a
                category
 - manifest.yaml: manifest file per release

This can be generated for everything, or only for a single (new)
release. The first is a rebuild, the second is for when a new
release is pushed to the CDN. In the second case the bare minimum
will be done to come to the same result as a full rebuild would
have done.

This requires read access to the S3 via the AWS API.

### CODE REVIEW

File: .github/workflows/update_cdn.yml [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        python -m pip install --upgrade pip
+        pip install -r cdn/requirements.txt
+
+    # By default, mime.types doesn't include one for YAML files. This results
+    # in "binary/octet-stream" as content-type on the CDN, which means
+    # browsers want to download the file instead of showing it inline. By
+    # adding this entry to mime.types, we fix that problem.
+    - name: Fix mimetypes
+      run: |
+        echo "text/vnd.yaml                                   yaml" | sudo tee -a /etc/mime.types
```
Comment (LordAro) [inline .github/workflows/update_cdn.yml]: I guess these extra spaces aren't *technically* necessary :)

File: .github/workflows/update_cdn.yml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          exit 1
+        fi
+
+    - name: Checkout
+      uses: actions/checkout@v2
+
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.7
```
Comment (LordAro) [inline .github/workflows/update_cdn.yml]: 3.8?

File: cdn/cdn_generator/common_files.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import shutil
+
+
+def generate_common_files():
+    pathlib.Path("generated").mkdir(parents=True, exist_ok=True)
+    shutil.copyfile("config.yaml", "generated/config.yaml")
+    shutil.copyfile("README.md", "generated/README.md")
+
+    pathlib.Path("generated/errors").mkdir(parents=True, exist_ok=True)
+    shutil.copyfile("errors/404.html", "generated/errors/404.html")
```
Comment (LordAro) [inline cdn/cdn_generator/common_files.py]: maybe a comment explaining why these nongenerated files need to be in the generated folder?

File: cdn/cdn_generator/directory_listing.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    pathlib.Path(f"generated/{prefix}").mkdir(parents=True, exist_ok=True)
+    with open(f"generated/{prefix}/index.html", "w") as f:
+        f.write("\n".join(html_content) + "\n")
+
+
+def _generate_directory_listing(prefix, files, folders, reverse_folders=False):
+    print(f"Creating directory listing for {prefix} ...")
+
+    # Use this to make sure the latest version is on top.
+    if reverse_folders:
```
Comment (LordAro) [inline cdn/cdn_generator/directory_listing.py]: this parameter feels a bit funny to me. The function isn't responsible for ensuring that `folders` is sorted, but is responsible for reversing it if requested? Either this should sort the folders, or reversing should be done outside this function

File: cdn/cdn_generator/helpers.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        folders.sort()
+
+    return files, folders
+
+
+def get_content(key):
+    try:
+        content = client.get_object(Bucket=BUCKET, Key=key,)
+    except ClientError:
+        print(f"Tried getting object '{key}'")
```
Comment (LordAro) [inline cdn/cdn_generator/helpers.py]: print to stderr?

File: cdn/cdn_generator/helpers.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            name = "testing"
+        else:
+            name = "stable"
+    elif override_name == "in-folder-name":
+        # These binaries are in the form of YYYYMMDD-NAME-...
+        name = folder.split("-")[1]
+    elif override_name == "nightly":
+        name = "nightly"
+    else:
+        raise Exception(f"Unknown override-name value '{override_name}' for {folder}")
```
Comment (LordAro) [inline cdn/cdn_generator/helpers.py]: KeyError? (or something other than generic Exception)

File: .github/workflows/testing.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  black:
+    name: Black
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Set up Python 3.7
+      uses: actions/setup-python@v1
+      with:
+        python-version: 3.7
```
[the remark this answers is not in the document]
Comment (TrueBrain) [reply .github/workflows/testing.yml]: I tested this on 3.6 and 3.7. I haven't on 3.8 yet, so if it is all the same to you, I keep it on 3.7 for now.

**Pull request merged**